src/dfi/validation.cpp

```diff
@@ -846,47 +848,118 @@ static void ProcessLoanEvents(const CBlockIndex *pindex, CCustomCSView &cache, c
-    if (pindex->nHeight < chainparams.GetConsensus().DF15FortCanningRoadHeight) {
+static void LiquidityForFuturesLimit(const CBlockIndex *pindex,
+                                     CCustomCSView &cache,
+                                     const CChainParams &chainparams,
+                                     const LoanTokenCollection &loanTokens,
+                                     const bool futureSwapBlock) {
+    if (pindex->nHeight < chainparams.GetConsensus().DF23Height) {
         return;
     }
 
     auto attributes = cache.GetAttributes();
 
-    CDataStructureV0 activeKey{AttributeTypes::Param, ParamIDs::DFIP2203, DFIPKeys::Active};
-    CDataStructureV0 blockKey{AttributeTypes::Param, ParamIDs::DFIP2203, DFIPKeys::BlockPeriod};
-    CDataStructureV0 rewardKey{AttributeTypes::Param, ParamIDs::DFIP2203, DFIPKeys::RewardPct};
-    if (!attributes->GetValue(activeKey, false) || !attributes->CheckKey(blockKey) ||
-        !attributes->CheckKey(rewardKey)) {
+    CDataStructureV0 activeKey{AttributeTypes::Param, ParamIDs::DFIP2211F, DFIPKeys::Active};
+    if (!attributes->GetValue(activeKey, false)) {
         return;
     }
 
-    CDataStructureV0 startKey{AttributeTypes::Param, ParamIDs::DFIP2203, DFIPKeys::StartBlock};
-    const auto startBlock = attributes->GetValue(startKey, CAmount{});
-    if (pindex->nHeight < startBlock) {
+    CDataStructureV0 blockKey{AttributeTypes::Param, ParamIDs::DFIP2211F, DFIPKeys::BlockPeriod};
+    const auto blockPeriod = attributes->GetValue(blockKey, DEFAULT_FS_LIQUIDITY_BLOCK_PERIOD);
+
+    const auto dusdToken = cache.GetToken("DUSD");
+    if (!dusdToken) {
         return;
     }
 
-    const auto blockPeriod = attributes->GetValue(blockKey, CAmount{});
-    if ((pindex->nHeight - startBlock) % blockPeriod != 0) {
+    const auto &dusdID = dusdToken->first;
+
+    std::set<DCT_ID> tokens;
+    for (const auto &[id, loanToken] : loanTokens) {
+        tokens.insert(id);
+    }
+
+    // Filter out DUSD
+    tokens.erase(dusdID);
+
+    // Store liquidity for loan tokens
+    cache.ForEachPoolPair([&](const DCT_ID &, const CPoolPair &poolPair) {
+        // Check for loan token
+        const auto tokenA = tokens.count(poolPair.idTokenA);
+        const auto tokenB = tokens.count(poolPair.idTokenB);
+        if (!tokenA && !tokenB) {
+            return true;
+        }
+
+        // Make sure this is the DUSD loan token pair
+        const auto dusdA = poolPair.idTokenA == dusdID;
+        const auto dusdB = poolPair.idTokenB == dusdID;
+        if (!dusdA && !dusdB) {
+            return true;
+        }
+
+        cache.SetLoanTokenLiquidityPerBlock(
+            {static_cast<uint32_t>(pindex->nHeight), poolPair.idTokenA.v, poolPair.idTokenB.v}, poolPair.reserveA);
+        cache.SetLoanTokenLiquidityPerBlock(
+            {static_cast<uint32_t>(pindex->nHeight), poolPair.idTokenB.v, poolPair.idTokenA.v}, poolPair.reserveB);
+
```

Comment:
@Bushstar: I just realized: we no longer need to remove dusd from the tokens (and even get the Id).
We now save loanTokenLiquidity in both directions, so no more need to check "token" vs "dusd". just check "both sides loantoken (inkl dusd)?"

so keep dusd in tokens and do
```
if(!tokens.count(poolPair.idTokenA) || !tokens.count(poolPair.idTokenB)) {
  return true;
}
// its a loanToken-loanToken pool -> save the liquidity of both sides
cache.SetLoan....
```

Reply:
I'd prefer to explicitly check that one pair is DUSD in case we add loan tokens pool pairs that do not include DUSD as these will not be used in future swaps.

Reply:
valid point. would not hurt (cause it would be just unnecessary data) to store it, but also doesn't hurt to check it here.